### PR TITLE
qemu.tests: Add case for qemu-img compare

### PR DESCRIPTION
**Pull request by pingl2016**
Compare images after convert qcow2 format to raw format

Signed-off-by: Ping Li <pingl@redhat.com>

ID: 1395550

### CODE REVIEW

**Comment by huangyum** [comment]
Is there any chance that you lost qemu_disk_img_compare.py ?

**Comment by pingl2016** [comment]
@huangyum Script  qemu_disk_img_convert.py is reused for this case. You can check "type" in configure file for refer.

File: qemu/tests/qemu_disk_img_convert.py [hunk 1]
```diff
@@ -33,6 +36,28 @@ def convert(self, t_params=None):
         self.trash.append(converted)
         return params
 
+    @error_context.context_aware
```
Comment (xutian) [inline qemu/tests/qemu_disk_img_convert.py]: not good to list compare mode here. better way is  compare mode in variants of configuration.

File: qemu/tests/qemu_disk_img_convert.py [hunk 2]
```diff
@@ -33,6 +36,28 @@ def convert(self, t_params=None):
...
+        error_context.context("Compare images with default and strict mode",
+                              logging.info)
+        for compare_mode in t_params.objects("compare_mode_list"):
+            if compare_mode == "strict":
+                t_params["compare_mode"] = compare_mode
+            image1 = self.image_filename
+            image2 = storage.get_image_filename(t_params, self.data_dir)
+            try:
+                super(ConvertTest, self).compare_images(t_params, image1, image2)
+                if compare_mode == "strict":
```
Comment (xutian) [inline qemu/tests/qemu_disk_img_convert.py]: not need to check compare module here, it should be check in super class compare function, here only need to know differ or same. if differ, log details.

File: qemu/tests/qemu_disk_img_convert.py (hunk 1 above)
Comment (pingl2016) [inline qemu/tests/qemu_disk_img_convert.py]: For default mode and strict mode, test steps are the same before comparing. After we simply specify "compare mode" in variants, It is a waste of time when both default mode and strict mode are tested. However, compare_mode_list can solve this issue by specify different mode like compare_mode_list="default", compare_mode_list="strict" or compare_mode_list="default strict"

**Comment by pingl2016** [comment]
Based on avocado-framework/avocado-vt#791

File: qemu/tests/qemu_disk_img_convert.py (hunk 2 above)
Comment (pingl2016) [inline qemu/tests/qemu_disk_img_convert.py]: For this case, compare should be failed when check in strict mode. So I think it makes sense to check.

File: qemu/tests/qemu_disk_img_convert.py [hunk 3]
```diff
@@ -33,6 +36,28 @@ def convert(self, t_params=None):
...
+                    raise error.TestFail("images are identical in strict mode")
+            except (exceptions.TestFail, exceptions.TestError), detail:
+                if compare_mode != "strict":
+                    raise error.TestFail("images are mismatch: %s" % detail)
+
```
Comment (PyLearner) [inline qemu/tests/qemu_disk_img_convert.py]: hi, the line 77 should be deleted "n_params = convert_test.create_snapshot()" , didn't see the steps in polarion case.

Comment (pingl2016) [inline qemu/tests/qemu_disk_img_convert.py]: It is used for another case qemu_disk_img.convert.snapshot_to_qcow2

File: qemu/tests/qemu_disk_img_convert.py (hunk 2 above)
Comment (xutian) [inline qemu/tests/qemu_disk_img_convert.py]: it expect results is differ,  please check in except section, to ignore the exception. like:
```
except (exceptions.TestFail, exceptions.TestError), detail:
      if compare_mode == "strict" and xxx:
          pass
```

File: qemu/tests/cfg/qemu_disk_img_compare.cfg [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+    start_vm = no
+    image_size = 20G
+    force_create_image = no
+    backup_image_before_testing = yes
+    restore_image_before_testing = yes
+    option_verified = "format"
+    image_convert = "image1"
+    convert_name_image1 = "images/image1_to_raw"
+    convert_format_image1 = "raw"
+    compare_mode_list = "False True"
```
Comment (suqinhuang) [inline qemu/tests/cfg/qemu_disk_img_compare.cfg]: True  False is not clear enough, suggest to update compare_mode_list to strict_mode

File: qemu/tests/qemu_disk_img_convert.py [hunk 5]
```diff
@@ -33,6 +36,28 @@ def convert(self, t_params=None):
...
+    def compare_test(self, t_params):
+        """
+        Compare images.
+
+        :param t_params: Dictionary with the test parameters
+        """
+        error_context.context("Compare images with default and strict mode",
+                              logging.info)
+        for strict_mode in t_params.objects("compare_mode_list"):
+            rv = None
```
Comment (suqinhuang) [inline qemu/tests/qemu_disk_img_convert.py]: 1. suggest to use more clearer name instead of rv.
2. move error_context.context here

**Comment by suqinhuang** [comment]
LGTM

**Comment by pingl2016** [comment]
(1/2) smp_8.8192m.repeat1.run_test.Host_RHEL.m7.u3.qcow2.virtio_blk.up.virtio_net.RHEL.7.3.x86_64.io-github-autotest-qemu.qemu_disk_img_compare: PASS (176.83 s)
 (2/2) smp_8.8192m.repeat1.run_test.Host_RHEL.m7.u3.qcow2.virtio_blk.up.virtio_net.Win7.x86_64.sp1.io-github-autotest-qemu.qemu_disk_img_compare: PASS (538.32 s)

File: qemu/tests/cfg/qemu_disk_img_compare.cfg [hunk 6]
```diff
@@ -0,0 +1,26 @@
+- qemu_disk_img_compare:
+    virt_test_type = qemu
+    type = qemu_disk_img_convert
+    kill_vm = yes
+    start_vm = no
```
Comment (pingl2016) [inline qemu/tests/cfg/qemu_disk_img_compare.cfg]: updated. thanks

File: qemu/tests/qemu_disk_img_convert.py (hunk 1 above)
Comment (xutian) [inline qemu/tests/qemu_disk_img_convert.py]: kill_vm = no  will help you to save time for test prepare steps.   variants will make your test matrix looks more clearly

Comment (pingl2016) [inline qemu/tests/qemu_disk_img_convert.py]: but i still need to convert images before comparing.

Comment (pingl2016) [inline qemu/tests/qemu_disk_img_convert.py]: When we test strict mode, we just use it to compare with normal mode. We don't have test scenario which only test strict mode.

**Comment by xutian** [comment]
looks Ok to me.

**Pull request merged**